src/main/java/minijava/ir/IREmitter.java

```diff
@@ -0,0 +1,769 @@
+  private static final Entity CALLOC;
+  private static final MethodType PRINT_INT_TYPE;
+  private static final Entity PRINT_INT;
+
+  static {
+    // If we consistently call InitFirm.init() throughout our code, we guarantee that
+    // Firm.init() will be called exactly once, even if e.g. the test suite also needs to
+    // call Firm.init().
+    InitFirm.init();
+    // Use 64bit pointers by default
+    // I don't see a reason we should, though.
+    //Mode.setDefaultModeP(Mode.createReferenceMode("_64bit", Mode.Arithmetic.TwosComplement, 64, 1));
+    INT_TYPE = new PrimitiveType(Mode.getIs());
+    BOOLEAN_TYPE = new PrimitiveType(Mode.getBu());
+    PTR_TYPE = new PrimitiveType(Mode.getP());
+    // We have to initialize these exactly once because of name clashes.
+    CALLOC_TYPE = new MethodType(new Type[] {PTR_TYPE, PTR_TYPE}, new Type[] {PTR_TYPE});
+    CALLOC = new Entity(Program.getGlobalType(), "calloc", CALLOC_TYPE);
+    PRINT_INT_TYPE = new MethodType(new Type[] {INT_TYPE}, new Type[] {});
+    PRINT_INT = new Entity(Program.getGlobalType(), "print_int", PRINT_INT_TYPE);
+  }
+
+  private final IdentityHashMap<Class, ClassType> classTypes = new IdentityHashMap<>();
+  private final IdentityHashMap<Method, Entity> methods = new IdentityHashMap<>();
+  private final IdentityHashMap<Field, Entity> fields = new IdentityHashMap<>();
+
+  /**
+   * Maps local variable definitions such as parameters and ... local variable definitions to their
+   * assigned index. Which is a firm thing.
+   */
+  private final IdentityHashMap<LocalVariable, Integer> localVarIndexes = new IdentityHashMap<>();
+  /** The Basic Block graph of the current function. */
+  private Graph graph;
+  /**
+   * Construction is a firm Node factory that makes sure that we don't duplicate expressions, thus
+   * making common sub expressions irrepresentable.
+   */
+  private Construction construction;
+  /**
+   * Stores the node's value in the current lvar (if there is any). This is crucial for assignment
+   * to work. E.g. in the expression x = 5, we analyze the variable expression x and get back its
+   * value, which is irrelevant for assignment, since we need the *address of* x. This is where we
+   * need storeInCurrentLval, which would store the expression node of the RHS (evaluating to 5) in
+   * the address of x.
+   *
+   * <p>Now, in an ideal world, this variable would be 'Node currentLval', but firm doesn't offer a
+   * function for getting the address of a local variable as a Node. So in order to not duplicate a
+   * lot of work (e.g. computing array offsets, etc.) in a mechanism without this variable, we
+   * abstract actual assignment out into a function.
+   */
+  @Nullable private Function<Node, Node> storeInCurrentLval;
+
+  public static void main(String[] main_args) {}
+
+  @Override
+  public Void visitProgram(minijava.ast.Program that) {
+    classTypes.clear();
+    methods.clear();
+    fields.clear();
```

Comment:
If you intend to make this Visitor reusable, you probably need to clear some more state. These things are easily forgotten, maybe it's better to just document that this visitor can only be used once?